torch/distributed/pipelining/_backward.py

```diff
@@ -303,19 +308,35 @@ def extract_tensors_with_grads(output_val, grad_val):
                 ), f"grad_value expected to have type {type(output_val)} but got {type(grad_val)}"
                 assert len(output_val) == len(grad_val)
                 for ov, gv in zip(output_val, grad_val):
-                    extract_tensors_with_grads(ov, gv)
+                    extract_tensors_with_grads(
+                        ov,
+                        gv,
+                        extract_tensors_with_grads,
+                    )
             elif isinstance(output_val, dict):
                 if grad_val is None:
                     return
                 assert isinstance(grad_val, dict)
                 assert set(output_val.keys()) == set(grad_val.keys())
                 for k in output_val.keys():
-                    extract_tensors_with_grads(output_val[k], grad_val[k])
+                    extract_tensors_with_grads(
+                        output_val[k], grad_val[k], extract_tensors_with_grads
+                    )
             else:
                 # Output is a non-tensor type; just ignore it
                 pass
 
-        extract_tensors_with_grads(stage_output, output_grads)
+        # Note: ref cycle
+        # break a ref cycle that would keep tensors alive until GC runs
+        # 1. extract_tensors_with_grads refers to a cell that holds refs to any vars defined in stage_backward
+        #    and used in extract_tensors_with_grads
+        # 2. extract_tensors_with_grads referred to both stage_output_tensors, output_grad_tensors,
+        #    and to itself (extract_tensors_with_grads) since it makes a recursive call
+        # 3. stage_output_tensors was kept alive by the above refcycle, and it holds activation tensors, which is bad
+        # fix -> explictly pass in the ref to the fn, so there is no gc cycle anymore
+        extract_tensors_with_grads(
```

Comment:
if you passed in `stage_output_tensors` and `output_grad_tensors` without the `extract_tensors_with_grads` ref, would that also fix the reference cycle? Since then `extract_tensors_with_grads` will no longer have to access variables in the outer scope so there should be no cycle, or is my understanding off.

Reply:
no, that would fix the tensor lifetime problem, which is almost as good.  However technically, there would still be a reference cycle between `extract_tensors_with_grads` and the cell.  This cycle would live on until GC kicked in.  So breaking the cycle is both cleaner (less vars passed in than passing in the two lists), and better (not only prevents leaking the tensors, but also cleans up the pyobjs earlier)

Reply:
Following up: can we move `extract_tensors_with_grads` out of the scope of `stage_backward`?

Reply:
yes, but, then we'd be forced to add the other 2 lists back as args to extract_tensors_with_grads, but we could delete the `extract_tensors_with_grads` argument since `extract_tensors_with_grads` is a global now instead of a cellvar.  If you guys want to make that change, its fine too.